### PR TITLE
Rewrite data download and tokenization pipeline

### PR DESCRIPTION
**Pull request by leogao2**
Now supports:

- Lmd based archive handling (this includes efficient zstd and tar handling)
- No more unpacking compressed files
- Multi-file support
- Vastly simplified dataset class

### CODE REVIEW

File: tools/corpora.py [hunk 1]
```diff
@@ -99,44 +93,24 @@ def vocab_file(self):
...
     def exists(self):
         """Checks if the dataset is present"""
         return os.path.isdir(f"{self.base_dir}/{self.name}")
 
     def download(self):
         """downloads dataset"""
-        os.makedirs(self.base_dir, exist_ok=True)
-        os.system(f"wget {self.url} -O {os.path.join(self.base_dir, os.path.basename(self.url))}")
+        for url in self.urls:
+            os.makedirs(self.base_dir, exist_ok=True)
```
Comment (sdtblck) [inline tools/corpora.py:103]: this can be moved outside the for loop

**Review by sdtblck — COMMENTED** [review]
lgtm! I'll test it out then merge

**Comment by sdtblck** [comment]
I'm getting this error when processing 'PileSubset' - I'm guessing it's encountering an empty document or something?

```python
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "tools/preprocess_data.py", line 98, in encode
    doc_ids[-1].append(Encoder.tokenizer.eod)
IndexError: list index out of range
"""
```

File: tools/corpora.py (hunk 1 above)
Comment (leogao2) [reply tools/corpora.py]: Fixed

**Comment by leogao2** [comment]
Added a filter for empty documents

**Comment by sdtblck** [comment]
Ok great, I'm going to add a few more pile datasets before we merge. Can lmd to tar.gz?

**Comment by leogao2** [comment]
It should be able to, and if there's any format it can't do that we need I'll update lmd to support it

**Comment by sdtblck** [comment]
Also, do we want to have the urls point to eaidata, or the eye?

**Review by sdtblck — APPROVED** [review]

**Pull request merged**